edited pack type tests

debonair opened the pull request:
closes #418

danfuterman commented:
@debonair - I see there has been some work done on this task - I'd like to clarify something now so that it's not an assumption. There should be backend validation to check the pack type when editing a donation (I can see some form of this being added in this PR), but what's also needed is a way to handle this in the frontend before form submission. This would be done by limiting the list of selectable pack types in the dropdown to those that do not produce components (i.e. disable those that cannot be selected), for the case where the donation is for an ineligible donor, and the donation batch is not historical.

i.e. this issue should be resolved in the frontend (to provide a good user experience), as well as in the backend (to ensure blood safety principles are upheld).

debonair commented:
@danfuterman please validate the approach I have taken with this task @bausmeier please review both the front and the backend code and @jonojhs you may test on my test box.

bausmeier commented on src/service/DonationCRUDService.java:
```diff
@@ -117,6 +117,18 @@ public Donation updateDonation(long donationId, DonationBackingForm donationBack
+        if (packType.getCountAsDonation() && !donorConstraintChecker.isDonorEligibleToDonate(donation.getDonor().getId())) {
+
+            DonationBatch donationBatch = donationBatchRepository.findDonationBatchByBatchNumber(
+                    donationBackingForm.getDonationBatchNumber());
```
Couldn't you just do `donation.getDonationBatch()` instead of using the repository?

bausmeier commented on test/factory/DonationViewModelFactoryTests.java:
```diff
@@ -31,15 +33,20 @@
         Long irrelevantAdverseEventId = 11L;
         AdverseEvent adverseEvent = anAdverseEvent().withId(irrelevantAdverseEventId).build();
         Donation donation = aDonation().withId(IRRELEVANT_DONATION_ID)
+                .withDonor(aDonor().withId(IRRELEVANT_DONATION_ID).build())
```
Donor shouldn't use the donation id.

debonair commented:
@bausmeier I have fixed the issues you highlighted

bausmeier commented on test/service/DonationCRUDServiceTests.java:
```diff
@@ -262,10 +273,15 @@ public void testUpdateDonation_shouldRetrieveAndUpdateDonation() {
+        when(packTypeRepository.getPackTypeById(IRRELEVANT_PACK_TYPE_ID)).thenReturn(irrelevantPackType);
+        when(donorConstraintChecker.isDonorEligibleToDonate(IRRELEVANT_DONOR_ID)).thenReturn(true);
+        when(donorRepository.findDonorById(IRRELEVANT_DONOR_ID)).thenReturn(expectedDonor);
+        when(donationBatchRepository.findDonationBatchByBatchNumber(donationBatchNumber)).thenReturn(donationBatch);
```
This line is no longer necessary.

bausmeier commented:
@debonair You didn't add a test for the new logic that you added. The code looks fine other than that.

danfuterman commented:
@debonair - Done some initial system review, but holding off from further review or feedback, pending the following information:
 - You have asked for validation on the approach you have taken, but can you describe this approach, so that this is explicit and clear before review starts?
 - Can you describe the use cases you have gone through yourself when testing this, to confirm that the functionality you have added is working correctly and as expected?
 - Can you provide example(s) to test on your dev server - it is expected that you have performed sanity checks once loading this onto your dev server. Providing info on the donors/donations/entities that were used for this will make it much quicker for anyone reviewing this to load up working examples to test on your dev. server.

This information will be expected in future prior to system review.

debonair commented:
@danfuterman I have added a canDonate  permission to the DonationViewModelFactory  which depends on whether the donor is deferred, i.e if the donor is deferred then canDonate is set to false.

On the frontend, I am using this flag to filter out packtypes that are counted as donations, i.e that produce components

I have also added both backend and frontend validation to warn the user and prevent any changes to pack types that may result in unsafe donations.

To test, I have added donations from both safe and unsafe donors to a donation batch and tested to see if the right pack type options appear in the drop down in both places where one can edit donations.

All this is on my Test box for functional review, I am currently adding tests to test the additional methods I wrote to achive the functionality i described above.

danfuterman commented:
@debonair - 
Have you also tested adding donations using pack types that do and don't produce components?
When I tested, the new rule was being applied to all donations, which meant that donations that did produce components were also limited to the pack types that could be selected when editing the donation.

The requirements suggested: "This would be done by limiting the list of selectable pack types in the dropdown to those that do not produce components (i.e. disable those that cannot be selected)"
When I tested, it looked like the non-selectable pack types had been removed/hidden from the list, rather than being disabled.

"All this is on my Test box for functional review, I am currently adding tests to test the additional methods I wrote to achive the functionality i described above."
This doesn't help with what was asked for:
"Providing info on the donors/donations/entities that were used for this will make it much quicker for anyone reviewing this to load up working examples to test on your dev. server."
Please provide info on specific donation DINs to test, or donation batches to use where you have done the sanity checks on your dev. server.

jonojhs commented:
@debonair please see feedback below - @danfuterman please correct me if I am wrong. 

-I am able to select any pack type when adding a donation for an inelible donor into a live donation batch (This should be limited to test only donations, only test only donations can be added into live donation batches for deferred/ineligible donors) 

-If I add a donation type that produces components and then add a deferral after the donation was captured I should still be able to select a different pack type that produces components - I now see this has been limited to pack types that don't produce components and upon clicking edit the pack type has dissapeared)
//note that pack type is single
![image](https://cloud.githubusercontent.com/assets/13638722/11563582/61bf3244-99dd-11e5-92da-a71e122dd945.png)
//Now I click edit
![image](https://cloud.githubusercontent.com/assets/13638722/11563589/6d09e45a-99dd-11e5-8ed4-f1c3cdcc6940.png)
//Dropdown displays pack types that do not produce components (donor was not deferred at the point of this donation)
![image](https://cloud.githubusercontent.com/assets/13638722/11563613/84f0f144-99dd-11e5-993e-8a6d3aa25431.png)


-Unsure what is happening here, the pack type is completely disabled - this is one of the donations you have added in an open donation batch - please let me know what you have done here. 
![image](https://cloud.githubusercontent.com/assets/13638722/11563404/6b27650a-99dc-11e5-9b07-791683310051.png)

debonair commented:
@danfuterman I have fixed the issues you highlighted and tested with Good Donor, who is not deferred, all options appear when you edit his donation and Testonly donor  some of whose options are disabled because he is deferred. I also had a meeting with @bausmeier @jonojhs  do discuss the scenarios around changing the pack types. The latest code is on my test box and am awaiting feedback from @jonojhs

danfuterman commented:
@debonair previous response still applies, and has not been addressed:

"Providing info on the donors/donations/entities that were used for this will make it much quicker for anyone reviewing this to load up working examples to test on your dev. server."
Please provide info on specific donation DINs to test, or donation batches to use where you have done the sanity checks on your dev. server.

jonojhs commented:
@debonair , thanks the following cases work

-Add a test only donation from a deferred donor into a historical donation batch - I can only change the pack type to pack types that don't produce components :+1: 

-Add a test only donation from a deferred donor into a live donation batch - I can only change the pack type to pack types that don't produce components :+1:  (this was the most critical use case)

//Current Issue
When adding a donation that produces components from a deferred donor into a historical donation batch, you cannot change the pack type at all - If the original pack type produced components I should be able to change it to any pack type.

![image](https://cloud.githubusercontent.com/assets/13638722/11583854/24a7df5c-9a67-11e5-9d2a-8e967e2a9b7c.png)
![image](https://cloud.githubusercontent.com/assets/13638722/11583841/0e3426e0-9a67-11e5-8e13-7a10e2fed318.png)

jonojhs commented:
@debonair - Issues incoming 

//Issue 1
I added a normal donor and added a donation into a historical donation batch, upon editing and saving the pack type and clicking done I get taken to the home screen 
![image](https://cloud.githubusercontent.com/assets/13638722/11584185/81c64ca2-9a6a-11e5-9806-baeaf1fa82e0.png)

//Issue 2
If I change the pack type from a component producing pack type to a non component producing packtype I cannot edit it again - only non component producing packtypes become editable

danfuterman commented:
@debonair - Jono's use cases all sound like cases you should have tested yourself before asking for review. It seems like you have not gone through this step:
 - Can you describe the use cases you have gone through yourself when testing this, to confirm that the functionality you have added is working correctly and as expected?

Further testing on hold until you're explicit about the use cases you have gone through.

debonair commented:
@danfuterman 

I took the following steps with @jonojhs 

1. Add donation from a normal donor to normal donationBatch then edit, no restrictions to pack type drop down
2. Add donation from a normal donor to historical donationBatch then edit, no restrictions to pack type drop down
3. Add donation from a deferred donor to normal donationBatch then edit,  restrictions apply to pack type drop down relevant error
4. Add donation from a deferred donor to historical donationBatch then edit, no restrictions to pack type drop down (since this is a historical batch)

normal donor is Good Donor (000005)
deferred donor is Testonly Donor (000006)

@bausmeier I made some changes and added tests since your last review. please take a look

jonojhs commented:
Sorry @debonair @danfuterman came across an interesting use case - Unsure whether it's worth negating/Just ignore as will depend on User.

Basically you can add as many donations per donor into a Live donation batch 
1- Add a normal donation
2- Change pack type to Test only
3- Add a normal donation
4- Change pack type to Test only
5- Add a normal donation
6- Change pack type to Test only
7- At this poiint you can now change all the pack types back to normal donations (expected behaviour)
![image](https://cloud.githubusercontent.com/assets/13638722/11589733/747b1900-9a93-11e5-8213-1b440ff58e24.png)


-------------------------------------
Referring to Issue1 I believe this is out of scope and has already been noted as a defect https://jembiprojects.jira.com/browse/BSIS-263

danfuterman commented:
@jonojhs thanks for picking up on that, I think it's best if we consider it out of scope for this issue.

jonojhs commented:
@debonair @danfuterman Unsure how best to tackle this (What the proposed workflow should be)

If you add a donation that produces components to normal donation batch - and then  add a deferral for whatevre reason and then try edit the pack type - you can only change the pack type to ones that do not produce components (the previous packtype becomes deselected upon edit too)

//Deferral
![image](https://cloud.githubusercontent.com/assets/13638722/11590643/33f8b3ea-9a9a-11e5-8d37-5bdf96c6b097.png)

//Edit Donation - Pack types not editable.
![image](https://cloud.githubusercontent.com/assets/13638722/11590648/3e6b46da-9a9a-11e5-80d2-d1b8f26fff0e.png)
![image](https://cloud.githubusercontent.com/assets/13638722/11590659/5117b5f2-9a9a-11e5-872a-7ac9958428ad.png)

danfuterman commented:
@jonojhs That use case sounds acceptable for now, i.e. not a priority for now.
The main use case we're working with for this:

- User completes form, submits. Realises they selected the wrong pack type, edits the form immediately after submitting and changes the pack type to the correct one, saves the form.

Feedback on the state of this task in consideration of that basic workflow would be greatly appreciated.

jonojhs commented:
@danfuterman In that case, I would say that this is done :+1: . Sorry @debonair for delays -  just trying to make sure I have covered all the bases 

Please note that this defect has not been covered in this task https://jembiprojects.jira.com/browse/BSIS-263

danfuterman commented:
@debonair - Done some initial system testing, seems to be working as expected :+1: 
Need to do some follow-up code review and further testing.

danfuterman commented:
@bausmeier could you follow up on, and prioritise review of this PR and https://github.com/jembi/bsis-frontend/pull/307.

bausmeier commented:
@danfuterman Okay I will. I skipped this one because of your [comment above](https://github.com/jembi/bsis/pull/432#issuecomment-161977515). What did you cover in the initial testing and what still needs to be tested?

danfuterman commented:
Thanks @bausmeier - I did some follow up system testing to check that the issues raised previously had been resolved. They appeared to have been resolved, but it'd be great to have some further testing done on this to test a few different scenarios and check the code to verify that all works as expected. Finally, it looks like the frontend PR has had some recent commits added that are not relevant and relate to enabling/disabling of pack types - I'm not sure if the follow-up commits removed these or not. Thanks!

bausmeier commented:
@danfuterman LGTM
